# Projet Traitement de données multidimensionnelles

Le projet vise à créer une intelligence artificielle (IA) utilisant un Large Language Model (LLM) pour catégoriser automatiquement les commentaires sur des séries télévisées. L'approche détaillée pour réaliser ce projet :

1. Choixd'un LLM adéquat
Il y a plusieursmodèles comme BERT, GPT-2 ..., chaque modèle a ses forces spécifiques pour différentes tâches de traitement de langage naturel.
Critères de Sélection : on doit prendre en compte les capacités de traitement de langage du modèle, la facilité d'utilisation, la disponibilité des ressources (comme les modèles pré-entraînés), et les exigences matérielles ( si un GPU est nécessaire), le domaine, la langue.
2. Évaluer les Performances du LLM sur un Ensemble de Test
Préparation des Données : on doit divisez nos 3 corpus en ensembles d''entraînement, de validation et de test.
Métriques de Performance : on peut utilisez des métriques telles que la précision, pour évaluer les performances initiales du LLM sur l'ensemble de test.
3. Optimiser les Performances
Ajustement de la Taille de l''Ensemble de Test : on doit testez différentes tailles pour l''ensemble de test et évaluez comment cela affecte les performances du modèle.
4. Fine-Tuning du LLM Sélectionné
Adaptation aux données Spécifiques : Fine-tunez le modèle sur notre ensemble d''entraînement pour améliorer ses performances sur les données de commentaires spécifiques.
Ajustements des Hyperparamètres : On peut expérimentez avec différents taux d''apprentissage, tailles de batch, et autres hyperparamètres.

In [1]:
import xml.etree.ElementTree as ET

# on lit les corpus 1, 2 et 3
chemin_fichier_xml1 = 'C:\\Users\\docme\\Desktop\\MASTER_2_CHPS\\IA_LLM_KAMEL\\PROJET_AMIRA_MEHRI_LLM\\corpus_1.xml'
chemin_fichier_xml2 = 'C:\\Users\\docme\\Desktop\\MASTER_2_CHPS\\IA_LLM_KAMEL\\PROJET_AMIRA_MEHRI_LLM\\corpus_2.xml'
chemin_fichier_xml3 = 'C:\\Users\\docme\\Desktop\\MASTER_2_CHPS\\IA_LLM_KAMEL\\PROJET_AMIRA_MEHRI_LLM\\corpus_3.xml'

# Parse les fichier XML
tree1 = ET.parse(chemin_fichier_xml1)
tree2 = ET.parse(chemin_fichier_xml2)
tree3 = ET.parse(chemin_fichier_xml3)
root1 = tree1.getroot()
root2 = tree2.getroot()
root3 = tree3.getroot()

# Fonction 1 pour extraire les informations
def extract_info(root1):
    # Extraire les informations de la série et des commentaires
    series_info1 = []
    for serie_commentaires in root1.findall('SERIE_COMMENTAIRES'):
        serie = serie_commentaires.find('SERIE').text
        serie_id = serie_commentaires.find('SERIE').get('id')
        comments = []
        for commentaire in serie_commentaires.findall('COMMENTAIRES/COMMENTAIRE'):
            comment_id = commentaire.get('id')
            author = commentaire.get('auth')
            evaluation = commentaire.get('eval')
            text = commentaire.text
            comments.append({
                'id': comment_id,
                'author': author,
                'evaluation': evaluation,
                'text': text
            })
        series_info1.append({
            'serie': serie,
            'serie_id': serie_id,
            'comments': comments
        })
    return series_info1

# Extraction des informations
series_info1 = extract_info(root1)
#print(series_info)  # Afficher les informations extraites
print(series_info1[:10])  # Affiche seulement les 100 premiers éléments

# Fonction 2 pour extraire les informations
def extract_info(root2):
    # Extraire les informations de la série et des commentaires
    series_info2 = []
    for serie_commentaires in root2.findall('SERIE_COMMENTAIRES'):
        serie = serie_commentaires.find('SERIE').text
        serie_id = serie_commentaires.find('SERIE').get('id')
        comments = []
        for commentaire in serie_commentaires.findall('COMMENTAIRES/COMMENTAIRE'):
            comment_id = commentaire.get('id')
            author = commentaire.get('auth')
            evaluation = commentaire.get('eval')
            text = commentaire.text
            comments.append({
                'id': comment_id,
                'author': author,
                'evaluation': evaluation,
                'text': text
            })
        series_info2.append({
            'serie': serie,
            'serie_id': serie_id,
            'comments': comments
        })
    return series_info2

# Extraction des informations
series_info2 = extract_info(root2)
#print(series_info2)  # Afficher les informations extraites
print(series_info2[:10])  # Affiche seulement les 100 premiers éléments

# Fonction 2 pour extraire les informations
def extract_info(root3):
    # Extraire les informations de la série et des commentaires
    series_info3 = []
    for serie_commentaires in root3.findall('SERIE_COMMENTAIRES'):
        serie = serie_commentaires.find('SERIE').text
        serie_id = serie_commentaires.find('SERIE').get('id')
        comments = []
        for commentaire in serie_commentaires.findall('COMMENTAIRES/COMMENTAIRE'):
            comment_id = commentaire.get('id')
            author = commentaire.get('auth')
            evaluation = commentaire.get('eval')
            text = commentaire.text
            comments.append({
                'id': comment_id,
                'author': author,
                'evaluation': evaluation,
                'text': text
            })
        series_info3.append({
            'serie': serie,
            'serie_id': serie_id,
            'comments': comments
        })
    return series_info3

# Extraction des informations
series_info3 = extract_info(root3)
#print(series_info2)  # Afficher les informations extraites
print(series_info3[:10])  # Affiche seulement les 100 premiers éléments



IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [ ]:
def score_to_stars(score):
    # Convertir le score en nombre flottant
    score = float(score.replace(',', '.'))
    # Attribuer des étoiles en fonction du score
    stars = '★' * int(score)  # Étoile pleine pour chaque point entier
    if score - int(score) >= 0.5:
        stars += '½'  # Ajouter une demi-étoile si nécessaire
    return stars

# Parcourir chaque série et chaque commentaire pour transformer les scores
for serie in series_info2:
    for comment in serie['comments']:
        score = comment['evaluation']
        stars = score_to_stars(score)
        print(f"Série: {serie['serie']}, Commentaire ID: {comment['id']}, Score: {score}, Évaluation en étoiles: {stars}")


In [2]:
from transformers import GPT2Tokenizer
import tensorflow as tf

# Initialiser le tokenizer
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')

# On Assure que le tokenizer utilise le même token que le modèle pour le padding
tokenizer.pad_token = tokenizer.eos_token

# Tokeniser et encoder tous les commentaires
encoded_comments = [tokenizer.encode_plus(
    comment['text'],
    truncation=True,
    padding='max_length',
    max_length=512,
    return_tensors="tf"
) for item in series_info2 for comment in item['comments']]

# Maintenant, on doit extraire les input_ids et attention_masks
input_ids = [ec['input_ids'][0] for ec in encoded_comments]
attention_masks = [ec['attention_mask'][0] for ec in encoded_comments]

# Cette opération suppose que encoded_comments est une liste de dictionnaires retournés par encode_plus
input_ids = tf.concat([ec['input_ids'] for ec in encoded_comments], 0)
attention_masks = tf.concat([ec['attention_mask'] for ec in encoded_comments], 0)


In [3]:
from sklearn.model_selection import train_test_split

# Convertission de nos évaluations de string à float et remplacez les virgules par des points
labels = tf.convert_to_tensor([float(comment['evaluation'].replace(',', '.')) for item in series_info2 for comment in item['comments']], dtype=tf.float32)
print(labels)


# Convertission de nos tenseurs en numpy arrays 
input_ids_np = input_ids.numpy()
attention_masks_np = attention_masks.numpy()
labels_np = labels.numpy()

# Division de nos données
train_inputs, test_inputs, train_labels, test_labels = train_test_split(input_ids_np, labels_np, random_state=42, test_size=0.1)
train_masks, test_masks, _, _ = train_test_split(attention_masks_np, labels_np, random_state=42, test_size=0.1)

tf.Tensor([4.5 3.  5.  ... 5.  4.5 4.5], shape=(7706,), dtype=float32)


In [4]:
# Vérifier la forme des données
print("Forme des inputs d'entraînement :", train_inputs.shape)
print("Forme des masques d'entraînement :", train_masks.shape)
print("Forme des étiquettes d'entraînement :", train_labels.shape)

# Vérifier que les tensors ne contiennent pas que des zéros, ce qui indiquerait un problème de padding
print("Inputs d'entraînement, premiers éléments :", train_inputs[0][:10])
print("Masques d'entraînement, premiers éléments :", train_masks[0][:10])


Forme des inputs d'entraînement : (6935, 512)
Forme des masques d'entraînement : (6935, 512)
Forme des étiquettes d'entraînement : (6935,)
Inputs d'entraînement, premiers éléments : [41840   235 41840   235 41840   235 41840   235 41840   235]
Masques d'entraînement, premiers éléments : [1 1 1 1 1 1 1 1 1 1]


In [5]:
from transformers import TFGPT2Model, GPT2Tokenizer
from tensorflow.keras.layers import Input, Dense, Dropout
from tensorflow.keras.models import Model
from tensorflow.keras.initializers import glorot_uniform
from tensorflow.keras.regularizers import l2  # Importer la régularisation L2

# Chargement du modèle GPT-2 pré-entraîné
gpt2_model = TFGPT2Model.from_pretrained('gpt2')

# Définition des couches d'entrée pour les IDs de tokens et les masques d'attention
input_ids_layer = Input(shape=(512,), dtype='int32', name='input_ids')
attention_masks_layer = Input(shape=(512,), dtype='int32', name='attention_mask')

# Obtontion des sorties du modèle GPT-2
outputs = gpt2_model(input_ids_layer, attention_mask=attention_masks_layer)

# GPT-2 n'a pas de 'pooler_output', donc nous utilisons directement les 'logits'
# qui sont les scores avant l'activation finale
# Prendre le dernier état caché pour représenter la séquence entière
sequence_output = outputs.last_hidden_state

# nous pouvons prendre le dernier token pour la classification
# L'indexation dépend de la direction de l'attention dans le modèle GPT-2 utilisé
clf_output = sequence_output[:, -1, :]  # Prendre le vecteur associé au dernier token

# Appliquer le dropout pour la régularisation
clf_output = Dropout(0.1)(clf_output)

# Ajouter une couche dense pour la classification finale avec régularisation L2
output_layer = Dense(1, activation='sigmoid', kernel_regularizer=l2(0.01))(clf_output)  # Pour la classification binaire

# Construction du le modèle
classification_model = Model(inputs=[input_ids_layer, attention_masks_layer], outputs=output_layer)

# Compilation du modèle
classification_model.compile(
    optimizer=tf.keras.optimizers.Adam(learning_rate=5e-5),
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

# Afficher le résumé du modèle
classification_model.summary()

All PyTorch model weights were used when initializing TFGPT2Model.

All the weights of TFGPT2Model were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFGPT2Model for predictions without further training.


Model: "model"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_ids (InputLayer)      [(None, 512)]                0         []                            
                                                                                                  
 attention_mask (InputLayer  [(None, 512)]                0         []                            
 )                                                                                                
                                                                                                  
 tfgpt2_model (TFGPT2Model)  TFBaseModelOutputWithPastA   1244398   ['input_ids[0][0]',           
                             ndCrossAttentions(last_hid   08         'attention_mask[0][0]']      
                             den_state=(None, 512, 768)                                       

Ce résumé nous indiques ces informations importantes : Entrée du Modèle:

* input_ids: Entrée pour les identifiants des tokens, avec une forme de (None, 512). Cela indique que le modèle attend une séquence de 512 tokens.
* attention_mask: Entrée pour les masques d'attention, également avec une forme de (None, 512).

* Layer GPT-2 (TFGPT2Model):

 - Le modèle GPT-2 est utilisé comme couche principale. Il prend input_ids et attention_mask comme entrées.
 - Cette couche produit un last_hidden_state de taille (None, 512, 768), où 512 est la longueur de la séquence et 768 est la taille des vecteurs de caractéristiques.

* Traitement Post-Modèle: 
       - tf.__operators__.getitem: Cette opération extrait le dernier état caché associé au dernier token (index -1) pour chaque séquence. La sortie de cette couche est de taille (None, 768).
       - dropout_37: Applique un dropout pour la régularisation, avec un taux de 0.1.

* Couche de Classification: Une couche dense avec une seule unité (None, 1) et une fonction d'activation 'sigmoid'. Cela suggère que votre modèle est configuré pour une classification binaire.
* Paramètres: - Le modèle a un total de 124,440,577 paramètres, tous étant entraînables. Cela reflète la taille et la complexité du modèle GPT-2.
              - Compilation: Le modèle est compilé avec une optimisation Adam et une perte de 'categorical_crossentropy'.

In [6]:
# Entraînement du modèle
# Ajuste `batch_size` et `epochs` selon mon pc  et les besoins du modèle
history = classification_model.fit(
    [train_inputs, train_masks],
    train_labels,
    validation_data=([test_inputs, test_masks], test_labels),
    batch_size=8,  # Taille du batch, peut être réduite ou augmentée selon la capacité de votre machine
    epochs=1,  # Nombre d'époques, à ajuster selon la convergence du modèle
    verbose=1  # Pour afficher la progression de l'entraînement, mettre à 2 pour moins de détails
)

# Évaluation du modèle sur l'ensemble de test
test_loss, test_accuracy = classification_model.evaluate(
    [test_inputs, test_masks],
    test_labels,
    verbose=1
)
print(f"Test Loss: {test_loss}, Test Accuracy: {test_accuracy}")


C:\Users\docme\anaconda3\envs\monPython\lib\site-packages\tensorflow\python\util\dispatch.py:1260: SyntaxWarning: In loss categorical_crossentropy, expected y_pred.shape to be (batch_size, num_classes) with num_classes > 1. Received: y_pred.shape=(None, 1). Consider using 'binary_crossentropy' if you only have 2 classes.
  return dispatch_target(*args, **kwargs)


25/25 [==============================] - 440s 18s/step - loss: 0.0641 - accuracy: 0.0246
Test Loss: 0.06412836909294128, Test Accuracy: 0.024643320590257645


In [11]:
# Fine-tune le modèle sur notre ensemble de données
history = classification_model.fit(
    [train_inputs, train_masks],  # Données d'entraînement
    train_labels,                 # Étiquettes d'entraînement
    validation_data=([test_inputs, test_masks], test_labels),  # Données de validation
    batch_size=16,                # on peut essayer d'ajuster la taille du batch
    epochs=2,                     # Et le nombre d'époques
    verbose=1                     # Pour afficher la progression de l'entraînement
)

# Sauvegarder le modèle après l'entraînement
classification_model.save('path_to_save_model')

# Évaluation du modèle sur l'ensemble de test
test_loss, test_accuracy = classification_model.evaluate(
    [test_inputs, test_masks],  # Données de test
    
    test_labels,                # Étiquettes de test
    verbose=1                   # Pour afficher les résultats de l'évaluation
)

print(f"Test Loss: {test_loss}, Test Accuracy: {test_accuracy}")

# nous pouvons également tracer l'historique de l'entraînement pour visualiser les performances du modèle au fil des époques
import matplotlib.pyplot as plt

plt.plot(history.history['accuracy'], label='accuracy')
plt.plot(history.history['val_accuracy'], label='val_accuracy')
plt.xlabel('Epoch')
plt.ylabel('Accuracy')
plt.ylim([0, 1])
plt.legend(loc='lower right')
plt.show()


Epoch 1/2
 58/434 [===>..........................] - ETA: 3:48:04 - loss: 0.0840 - accuracy: 0.0216

KeyboardInterrupt: 

Mon PC n'a pas pu terminer la phase de fine tuning, il est même bloqué donc j'ai arrêté l'exécution du noyau.

J''ai choisis le modèle GPT2 car il est puissant il travaille sur plusieurs domaines, il fonctionne de manière auto-régressive, générant un token à la fois et utilisant les tokens précédemment générés comme contexte pour la génération du token suivant. Aussi GPT se base exclusivement sur la partie décodeur du Transformer.

J'' ai essayé un seul corpus, car je n''ai pas un ordinateur plus puisssaant pour l''entrainement des modèles, j''ai créé la fonction score_to_stars qui convertit les commentaires en des étoiles mais j''ai eu le temps de l'utiliser dans mon modèle.   